js/userDatabase.js

```diff
@@ -0,0 +1,118 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+(function(exports) {
+  const DATABASE_VERSION = 1;
+  const DATABASE_NAME = "NewTabData";
+
+  const OBJECT_STORE_PREFS = "prefs";
+  const PINNED_LINKS_PREF = "pinnedLinks";
+
+  const gUserDatabase = {
+    _database: null,
+
+    init(mockDB) {
+      return new Promise((resolve, reject) => {
+        var db = mockDB || window.indexedDB;
+        var request = db.open(DATABASE_NAME, DATABASE_VERSION);
+
+        request.onerror = event => {
+          gUserDatabase._logError(event, "Cannot open an indexedDB connection");
+          reject(new Error(event.target.errorCode));
+        };
+        request.onsuccess = event => {
+          gUserDatabase._onDatabaseOpenSuccess(event).then(resolve);
+        };
+        request.onupgradeneeded = event => {
+          gUserDatabase._onDatabaseUpgrade(event);
```

Comment:
_onDatabaseUpgrade should probably return a promise, which is then used to `resove` this promise. Otherwise, init will never resolve is there is an upgrade needed.